### PR TITLE
feat: sync delete mode

### PR DESCRIPTION
**Pull request by elrrrrrrr**
> 为了避免部分 npm 包误封、误删，导致生产环境影响，新增 syncDeleteMode 配置，允许自定义同步策略

* 新增 `syncDeleteMode` : 'ignore' | 'block' | 'delete'
  * delete: 目前默认值，同步删包事件
  * ignore: 忽略 upstream 所有删包事件
  * block: 不做物理删除，只新增 block 记录，不允许访问，除非管理员手动恢复并更新 `syncPackageBlockList`
* `npm-security-holder` 场景也判断为删包事件
* 更新原有删包流程，统一处理，调整部分日志输出

---------------

> New `syncDeleteMode` to allow custom syncing policy to avoid some npm packages being blocked or deleted by mistake.

* Add `syncDeleteMode` : 'ignore' | 'block' | 'delete'
  * delete: by default, sync delete events
  * ignore: ignore all upstream delete events
  * block: only add block records, cant access unless the administrator manually restores and update `syncPackageBlockList`.
* `npm-security-holder` event is also determined to be a delete event
* Update the original packet deletion process, update log output by the way

Translated with www.DeepL.com/Translator (free version)

### CODE REVIEW

File: config/config.default.ts [hunk 1]
```diff
@@ -22,7 +23,8 @@ export default (appInfo: EggAppConfig) => {
     //  - none: don't sync npm package, just redirect it to sourceRegistry
     //  - all: sync all npm packages
     //  - exist: only sync exist packages, effected when `enableCheckRecentlyUpdated` or `enableChangesStream` is enabled
-    syncMode: 'none',
+    syncMode: SyncMode.none,
```
Comment (fengmk2) [inline config/config.default.ts:26]: 这个得兼容一下原来的 3个字符串参数定义，否则已经部署的私有化应用会不会有问题？

File: app/core/service/PackageSyncerService.ts [hunk 2]
```diff
@@ -209,6 +219,84 @@ export class PackageSyncerService extends AbstractService {
...
+  // - delete: remove the package from local registry
+  // - block: block the package, update the manifest.block, instead of delete versions
+  // 根据 syncDeleteMode 配置，处理删包场景
+  // - ignore: 不做任何处理，直接结束任务
+  // - delete: 删除包数据，包括 manifest 存储
+  // - block: 软删除 将包标记为 block，用户无法直接使用
+  private async syncDeletePkg({ task, pkg, logUrl, url, logs, data }: syncDeletePkgOptions) {
+    const fullname = task.targetName;
+    const failEnd = `❌❌❌❌❌ ${url || fullname} ❌❌❌❌❌`;
+    const { syncDeleteMode = SyncDeleteMode } = this.config.cnpmcore;
```
Comment (fengmk2) [inline app/core/service/PackageSyncerService.ts:260]: syncDeleteMode = SyncDeleteMode 是什么语法？

File: app/core/service/PackageSyncerService.ts [hunk 3]
```diff
@@ -209,6 +219,84 @@ export class PackageSyncerService extends AbstractService {
...
+      return;
+    }
+
+    // ignore deleted package
+    if (syncDeleteMode === SyncDeleteMode.ignore) {
+      logs.push(`[${isoNow()}] 🟢 Skip remove since config.syncDeleteMode = ignore`);
+    }
+
+    // block deleted package
+    if (syncDeleteMode === SyncDeleteMode.block) {
```
Comment (fengmk2) [inline app/core/service/PackageSyncerService.ts:280]: 用 else if？否则每次都执行3个 if 逻辑判断

File: app/core/service/PackageSyncerService.ts [hunk 4]
```diff
@@ -209,6 +219,84 @@ export class PackageSyncerService extends AbstractService {
...
+    }
+
+    // unpublished
+    const timeMap = data.time || {};
+    if (timeMap.unpublished) {
+      return true;
+    }
+
+    // security holder
+    if (data?.repository === 'npm/security-holder') {
```
[the remark this answers is not in the document]
Comment (fengmk2) [reply app/core/service/PackageSyncerService.ts:242]: 加上 maintainers 里面包含 npm 做2次确认，这样应该就不会误报了。

File: config/config.default.ts (hunk 1 above)
Comment (elrrrrrrr) [reply config/config.default.ts:26]: enum 最后还是编译成 string，原来的配置能兼容的

File: app/core/service/PackageSyncerService.ts (hunk 2 above)
Comment (elrrrrrrr) [reply app/core/service/PackageSyncerService.ts]: 😂 是错误的语法

File: app/core/service/PackageSyncerService.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (elrrrrrrr) [reply app/core/service/PackageSyncerService.ts:242]: security-holder 的时候 maintainers 里也是空数组

我们改成判断 versions 内版本符合 ^0.0.1-security
且 _npmUser.name = 'npm' ?

**Comment by fengmk2** [comment]
@elrrrrrrr CI 挂了。

**Review by fengmk2 — APPROVED** [review]

**Pull request merged**